File: 0015-3sum/three-pointer.cs
using System;
using System.Collections;

public class Solution
{
    public IList<IList<int>> ThreeSum(int[] nums)
    {
        int size = nums.Length;

        // Sort array
        Array.Sort(nums);

        // Initialize return IList
        IList<IList<int>> result = new List<IList<int>>();

        for (int i = 0; i < (nums.Length - 2); i++)
        {
            // Skip duplicates
            if ((i > 0) && (nums[i] == nums[i - 1]))
            {
                continue;
            }
            else
            {
                int j = i + 1;
                int k = nums.Length - 1;

                while (j < k)
                {
                    int sum = nums[i] + nums[j] + nums[k];

                    if (sum == 0)
                    {
                        // Add triplet
                        result.Add(new List<int> {nums[i], nums[j], nums[k]});
                        k--;

                        // Skip duplicates
                        while ((j < k) && (nums[k] == nums[k + 1]))
                        {
                            k--;
                        }
                    }
                    else if (sum > 0)
                    {
                        k--;
                    }
                    else
                    {
                        j++;
                    }
                }
            }
        }

        return result;
    }
}

File: 0015-3sum/three-pointer.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        size = len(nums)

        # Sort array
        nums.sort()

        # Initialize array of arrays
        result = []

        for i in range(size - 2):
            # Skip duplicates
            if ((i > 0) and (nums[i] == nums[i - 1])):
                continue
            else:
                j = i + 1
                k = size - 1

                while (j < k):
                    sum = nums[i] + nums[j] + nums[k]

                    if (sum == 0):
                        # Add triplet
                        result.append([nums[i], nums[j], nums[k]])

                        k -= 1

                        # Skip duplicates
                        while ((j < k) and (nums[k] == nums[k + 1])):
                            k -= 1
                    elif (sum > 0):
                        k -= 1
                    else:
                        j += 1
        return result